- Target variable - Price
- Regression problem

# **Import Libraries**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# **Load the Data**

In [3]:
df = pd.read_csv("second_hand_cars.csv")

In [4]:
df.head()

,Company Name,Car Name,Variant,Fuel Type,Tyre Condition,Make Year,Owner Type,Registration Number,Mileage,Price,Transmission Type,Body Color,Service Record,Insurance,Registration Certificate,Accessories
0,Maruti Suzuki,Cruze,EX,CNG,Needs Replacement,2018,Second,84-436-5584,52798,759107,Manual,Grey,Major Service at 50418 km,No Current Insurance,Not Available,"Music System, Sunroof, Alloy Wheels"
1,Kia,Seltos,RXE,Petrol,New,2020,Third,79-114-3166,43412,505071,Automatic,Maroon,Major Service at 131313 km,No Current Insurance,Available,NaN
2,Kia,Accord,RXE,Petrol,New,2022,Second,41-358-3344,95219,635322,Automatic (Tiptronic),Black,No Service Record,No Current Insurance,Available,NaN
3,Nissan,Seltos,Highline,Diesel,Used,2024,Third,92-708-1763,70370,483152,Automatic (Tiptronic),Maroon,Major Service at 98115 km,Valid Until [date],Available,"Music System, Alloy Wheels"
4,Chevrolet,Kwid,Highline,Petrol,Used,2018,Second,76-154-5485,85852,712961,Automatic (Tiptronic),Silver,Major Service at 135665 km,No Current Insurance,Not Available,"GPS, Music System"


# **Data Cleaning and Preprocessing**

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Company Name              2500 non-null   str  
 1   Car Name                  2500 non-null   str  
 2   Variant                   2238 non-null   str  
 3   Fuel Type                 2500 non-null   str  
 4   Tyre Condition            2500 non-null   str  
 5   Make Year                 2500 non-null   int64
 6   Owner Type                2500 non-null   str  
 7   Registration Number       2500 non-null   str  
 8   Mileage                   2500 non-null   int64
 9   Price                     2500 non-null   int64
 10  Transmission Type         2500 non-null   str  
 11  Body Color                2500 non-null   str  
 12  Service Record            2500 non-null   str  
 13  Insurance                 2500 non-null   str  
 14  Registration Certificate  2500 non-null   str  
 15

In [6]:
df.isnull().sum()

Company Name                  0
Car Name                      0
Variant                     262
Fuel Type                     0
Tyre Condition                0
Make Year                     0
Owner Type                    0
Registration Number           0
Mileage                       0
Price                         0
Transmission Type             0
Body Color                    0
Service Record                0
Insurance                     0
Registration Certificate      0
Accessories                 482
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.describe()

,Make Year,Mileage,Price
count,2500.000000,2500.000000,2500.000000
mean,2019.521600,104777.792000,608120.897600
std,2.894146,55544.487467,231056.126713
min,2015.000000,10010.000000,200176.000000
25%,2017.000000,56313.000000,407791.000000
50%,2020.000000,104209.500000,612012.500000
75%,2022.000000,152149.250000,804567.000000
max,2024.000000,199755.000000,999826.000000


In [15]:
# check if missing values are concentrated in particular cars
df[df['Variant'].isnull()]['Car Name'].value_counts()

Car Name
Elantra    37
Camry      30
Accord     29
Kwid       29
Seltos     28
Polo       26
F-150      23
Swift      23
Sunny      20
Cruze      17
Name: count, dtype: int64

In [16]:
df[df['Accessories'].isnull()]['Car Name'].value_counts()

Car Name
Elantra    57
Sunny      54
Polo       54
Accord     51
F-150      47
Camry      47
Swift      47
Seltos     45
Cruze      41
Kwid       39
Name: count, dtype: int64

In [17]:
df[df['Variant'].isnull()][
    ['Company Name', 'Car Name', 'Variant', 'Fuel Type', 'Make Year',
     'Owner Type', 'Mileage', 'Price', 'Accessories']
]

,Company Name,Car Name,Variant,Fuel Type,Make Year,Owner Type,Mileage,Price,Accessories
6,Maruti Suzuki,Camry,NaN,CNG,2017,Third,70676,550187,GPS
7,Toyota,F-150,NaN,CNG,2021,Second,197377,344563,"GPS, Leather Seats"
10,Chevrolet,Sunny,NaN,CNG,2019,Third,56900,398364,"Leather Seats, Music System, Alloy Wheels, GPS"
13,Chevrolet,Camry,NaN,Petrol,2019,Third,91642,398425,"Sunroof, GPS"
22,Kia,Accord,NaN,Diesel,2019,Third,181920,232657,"Leather Seats, GPS, Alloy Wheels, Sunroof"
...,...,...,...,...,...,...,...,...,...
2456,Honda,Cruze,NaN,Petrol,2020,Second,189953,737067,"Alloy Wheels, Music System"
2466,Toyota,Kwid,NaN,Petrol,2022,Second,105646,423827,Music System
2467,Chevrolet,Cruze,NaN,CNG,2022,Third,68893,696446,"Music System, Sunroof"
2476,Hyundai,Kwid,NaN,CNG,2018,Second,128247,978228,"Music System, Sunroof"


In [18]:
df[df['Accessories'].isnull()][
    ['Company Name', 'Car Name', 'Variant', 'Fuel Type', 'Make Year',
     'Owner Type', 'Mileage', 'Price', 'Accessories']
]

,Company Name,Car Name,Variant,Fuel Type,Make Year,Owner Type,Mileage,Price,Accessories
1,Kia,Seltos,RXE,Petrol,2020,Third,43412,505071,NaN
2,Kia,Accord,RXE,Petrol,2022,Second,95219,635322,NaN
27,Ford,Accord,NaN,Petrol,2024,Second,176050,649268,NaN
30,Volkswagen,Sunny,XL,CNG,2018,Second,52809,565267,NaN
37,Volkswagen,Elantra,NaN,Petrol,2020,Second,38775,386004,NaN
...,...,...,...,...,...,...,...,...,...
2468,Renault,Swift,Highline,Diesel,2023,First,153838,852736,NaN
2470,Ford,F-150,RXE,CNG,2022,First,94065,350918,NaN
2484,Ford,Polo,EX,CNG,2019,Second,144772,577499,NaN
2488,Chevrolet,Accord,XL,Diesel,2019,Second,95168,708314,NaN


In [19]:
# view both variant and accessories together

df[df['Variant'].isnull() | df['Accessories'].isnull()][
    ['Company Name', 'Car Name', 'Variant', 'Fuel Type', 'Make Year',
     'Owner Type', 'Mileage', 'Price', 'Accessories']
]

,Company Name,Car Name,Variant,Fuel Type,Make Year,Owner Type,Mileage,Price,Accessories
1,Kia,Seltos,RXE,Petrol,2020,Third,43412,505071,NaN
2,Kia,Accord,RXE,Petrol,2022,Second,95219,635322,NaN
6,Maruti Suzuki,Camry,NaN,CNG,2017,Third,70676,550187,GPS
7,Toyota,F-150,NaN,CNG,2021,Second,197377,344563,"GPS, Leather Seats"
10,Chevrolet,Sunny,NaN,CNG,2019,Third,56900,398364,"Leather Seats, Music System, Alloy Wheels, GPS"
...,...,...,...,...,...,...,...,...,...
2476,Hyundai,Kwid,NaN,CNG,2018,Second,128247,978228,"Music System, Sunroof"
2484,Ford,Polo,EX,CNG,2019,Second,144772,577499,NaN
2488,Chevrolet,Accord,XL,Diesel,2019,Second,95168,708314,NaN
2490,Toyota,Camry,ZXI,CNG,2024,Third,170670,952572,NaN


In [22]:
print("Variant missing:", df['Variant'].isnull().sum())
print("Accessories missing:", df['Accessories'].isnull().sum())

print("Variant missing %:", df['Variant'].isnull().mean() * 100)
print("Accessories missing %:", df['Accessories'].isnull().mean() * 100)

Variant missing: 262
Accessories missing: 482
Variant missing %: 10.48
Accessories missing %: 19.28


- Variant has moderate missingness, with 262 missing values (10.48%).
- Accessories has relatively high missingness, with 482 missing values (19.28%).
- Rather than removing the affected observations, categorical missing values were replaced with "Unknown". This retained the available observations while avoiding potentially inaccurate assumptions about the missing values.
- Categorical imputation using "Unknown" was preferred because "Unknown" becomes a separate category that the model can potentially learn from.
- Mode imputation was not used, as assigning the most frequent category could be misleading. The variant cannot reliably be inferred from Company Name + Car Name in this dataset.

In [23]:
# filling variant NaN values with "unknown"
df['Variant'] = df['Variant'].fillna('Unknown')

In [24]:
# filling Accessories NaN values with "unknown"
df['Accessories'] = df['Accessories'].fillna('Unknown')

In [ ]:
# confirming if NaN values have been filled
df[['Variant', 'Accessories']].isnull().sum()

Variant        0
Accessories    0
dtype: int64

In [26]:
print(df['Variant'].value_counts())
print(df['Accessories'].value_counts())

Variant
LE          270
Unknown     262
GT          261
RXE         255
Highline    254
SE          250
ZXI         245
EX          241
XL          234
SL          228
Name: count, dtype: int64
Accessories
Unknown                                               482
Sunroof                                               119
GPS                                                    96
Music System                                           96
Leather Seats                                          94
                                                     ... 
GPS, Alloy Wheels, Sunroof, Leather Seats               1
Sunroof, Music System, GPS, Leather Seats               1
Music System, Sunroof, Alloy Wheels, Leather Seats      1
Sunroof, Alloy Wheels, GPS, Music System                1
Alloy Wheels, GPS, Leather Seats, Music System          1
Name: count, Length: 204, dtype: int64


- The most common actual variant is LE with 270 observations, while there are 262 missing values. If you filled all 262 missing values with LE, you'd artificially increase LE from 270 → 532, substantially changing the distribution.
- Using mode to impute accessories one would essentially assume that every missing entry had whatever the most common accessory combination is and that is not a very defensible assumption.

#### Because Accessories contains combinations of several accessories, it will be split into binary columns.

In [27]:
# create binary columns for each accessory
accessories = [
    'GPS',
    'Sunroof',
    'Music System',
    'Leather Seats',
    'Alloy Wheels'
]

for accessory in accessories:
    df[f'Has_{accessory.replace(" ", "_")}'] = (
        df['Accessories'].str.contains(accessory, regex=False)
    ).astype(int)

In [28]:
# check results 
df[
    ['Accessories',
     'Has_GPS',
     'Has_Sunroof',
     'Has_Music_System',
     'Has_Leather_Seats',
     'Has_Alloy_Wheels']
].head(10)

,Accessories,Has_GPS,Has_Sunroof,Has_Music_System,Has_Leather_Seats,Has_Alloy_Wheels
0,"Music System, Sunroof, Alloy Wheels",0,1,1,0,1
1,Unknown,0,0,0,0,0
2,Unknown,0,0,0,0,0
3,"Music System, Alloy Wheels",0,0,1,0,1
4,"GPS, Music System",1,0,1,0,0
5,"Sunroof, Music System, Leather Seats",0,1,1,1,0
6,GPS,1,0,0,0,0
7,"GPS, Leather Seats",1,0,0,1,0
8,"GPS, Music System, Sunroof, Leather Seats",1,1,1,1,0
9,"Music System, Leather Seats, Alloy Wheels",0,0,1,1,1


- 1 means the car has that accessory and 0 means it doesn't. Unknown will result in all five being 0.

In [29]:
# remove original accessories column
df.drop('Accessories', axis=1, inplace=True)

In [31]:
df.head(30)

,Company Name,Car Name,Variant,Fuel Type,Tyre Condition,Make Year,Owner Type,Registration Number,Mileage,Price,Transmission Type,Body Color,Service Record,Insurance,Registration Certificate,Has_GPS,Has_Sunroof,Has_Music_System,Has_Leather_Seats,Has_Alloy_Wheels
0,Maruti Suzuki,Cruze,EX,CNG,Needs Replacement,2018,Second,84-436-5584,52798,759107,Manual,Grey,Major Service at 50418 km,No Current Insurance,Not Available,0,1,1,0,1
1,Kia,Seltos,RXE,Petrol,New,2020,Third,79-114-3166,43412,505071,Automatic,Maroon,Major Service at 131313 km,No Current Insurance,Available,0,0,0,0,0
2,Kia,Accord,RXE,Petrol,New,2022,Second,41-358-3344,95219,635322,Automatic (Tiptronic),Black,No Service Record,No Current Insurance,Available,0,0,0,0,0
3,Nissan,Seltos,Highline,Diesel,Used,2024,Third,92-708-1763,70370,483152,Automatic (Tiptronic),Maroon,Major Service at 98115 km,Valid Until [date],Available,0,0,1,0,1
4,Chevrolet,Kwid,Highline,Petrol,Used,2018,Second,76-154-5485,85852,712961,Automatic (Tiptronic),Silver,Major Service at 135665 km,No Current Insurance,Not Available,1,0,1,0,0
5,Volkswagen,Cruze,LE,Petrol,New,2021,Second,21-711-8793,28516,209930,Automatic,Gold,Major Service at 63857 km,No Current Insurance,Available,0,1,1,1,0
6,Maruti Suzuki,Camry,Unknown,CNG,Used,2017,Third,92-504-4180,70676,550187,Automatic,Red,Full Service History,Valid Until [date],Not Available,1,0,0,0,0
7,Toyota,F-150,Unknown,CNG,New,2021,Second,78-995-2467,197377,344563,Automatic,Black,No Service Record,Valid Until [date],Not Available,1,0,0,1,0
8,Volkswagen,Kwid,ZXI,Petrol,New,2022,Third,87-719-2138,155320,596554,Manual,Grey,Full Service History,No Current Insurance,Available,1,1,1,1,0
9,Ford,Swift,GT,Petrol,Used,2018,Second,41-421-7619,16544,387560,Manual,White,Full Service History,Valid Until [date],Not Available,0,0,1,1,1


In [32]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Company Name              2500 non-null   str  
 1   Car Name                  2500 non-null   str  
 2   Variant                   2500 non-null   str  
 3   Fuel Type                 2500 non-null   str  
 4   Tyre Condition            2500 non-null   str  
 5   Make Year                 2500 non-null   int64
 6   Owner Type                2500 non-null   str  
 7   Registration Number       2500 non-null   str  
 8   Mileage                   2500 non-null   int64
 9   Price                     2500 non-null   int64
 10  Transmission Type         2500 non-null   str  
 11  Body Color                2500 non-null   str  
 12  Service Record            2500 non-null   str  
 13  Insurance                 2500 non-null   str  
 14  Registration Certificate  2500 non-null   str  
 15

# **Feature Engineering**

Did not create price per mile as a feature since price is the target variable and it would cause data leakage since it would mean using the target variable to create one of the predictions.

In [35]:
# car age
df['Car_Age'] = 2026 - df['Make Year']

# price per mile
df['Price per Mile'] = df['Price'] / df['Mileage'].replace(0, 1) #to avoid division by 0

In [37]:
# check new features
df[['Make Year', 'Car_Age', 'Price', 'Mileage', 'Price per Mile']].head()

,Make Year,Car_Age,Price,Mileage,Price per Mile
0,2018,8,759107,52798,14.377571
1,2020,6,505071,43412,11.634364
2,2022,4,635322,95219,6.672219
3,2024,2,483152,70370,6.865880
4,2018,8,712961,85852,8.304536


# **Encoding**

- Used one-hot encoding

In [38]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Company Name              2500 non-null   str    
 1   Car Name                  2500 non-null   str    
 2   Variant                   2500 non-null   str    
 3   Fuel Type                 2500 non-null   str    
 4   Tyre Condition            2500 non-null   str    
 5   Make Year                 2500 non-null   int64  
 6   Owner Type                2500 non-null   str    
 7   Registration Number       2500 non-null   str    
 8   Mileage                   2500 non-null   int64  
 9   Price                     2500 non-null   int64  
 10  Transmission Type         2500 non-null   str    
 11  Body Color                2500 non-null   str    
 12  Service Record            2500 non-null   str    
 13  Insurance                 2500 non-null   str    
 14  Registration Certif

In [39]:
from sklearn.preprocessing import OneHotEncoder

categorical_cols = [
    'Company Name',
    'Car Name',
    'Variant',
    'Fuel Type',
    'Tyre Condition',
    'Owner Type',
    'Transmission Type',
    'Body Color',
    'Service Record',
    'Insurance',
    'Registration Certificate'
]

encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

encoded_data = encoder.fit_transform(df[categorical_cols])

encoded_df = pd.DataFrame(
    encoded_data,
    columns=encoder.get_feature_names_out(categorical_cols),
    index=df.index
)

In [43]:
# combine encoded columns with numerical features
numeric_cols = [
    'Make Year',
    'Mileage',
    'Car_Age',
    'Price per Mile',
    'Has_GPS',
    'Has_Sunroof',
    'Has_Music_System',
    'Has_Leather_Seats',
    'Has_Alloy_Wheels'
]

df_encoded = pd.concat(
    [df[numeric_cols], encoded_df, df[['Price']]],
    axis=1
)

In [44]:
df_encoded.head()

,Make Year,Mileage,Car_Age,Price per Mile,Has_GPS,Has_Sunroof,Has_Music_System,Has_Leather_Seats,Has_Alloy_Wheels,Company Name_Chevrolet,...,Service Record_Major Service at 99208 km,Service Record_Major Service at 99441 km,Service Record_Major Service at 99840 km,Service Record_Major Service at 99928 km,Service Record_No Service Record,Insurance_No Current Insurance,Insurance_Valid Until [date],Registration Certificate_Available,Registration Certificate_Not Available,Price
0,2018,52798,8,14.377571,0,1,1,0,1,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,759107
1,2020,43412,6,11.634364,0,0,0,0,0,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,505071
2,2022,95219,4,6.672219,0,0,0,0,0,0.0,...,0.0,0.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,635322
3,2024,70370,2,6.865880,0,0,1,0,1,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,483152
4,2018,85852,8,8.304536,1,0,1,0,0,1.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,712961


# Split data to train/test set

In [45]:
from sklearn.model_selection import train_test_split

X = df_encoded.drop('Price', axis=1)
y = df_encoded['Price']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# **Feature Scaling**

used standard scaler since it puts the numerical features on a comparable scale without unnecessarily restricting their range.

In [46]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [48]:
# convert back to dataframe
X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns,
    index=X_train.index
)

X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns,
    index=X_test.index
)

# **Dimensionalty Reduction**

### 5.1. Correlation Analysis

In [53]:
corr_matrix = X_train_scaled.corr()

In [54]:
# find features with correlation above 0.8
upper = corr_matrix.abs().where(
    np.triu(np.ones(corr_matrix.shape), k=1).astype(bool)
)

high_corr = [
    column for column in upper.columns
    if any(upper[column] > 0.8)
]

print("Highly correlated features:", high_corr)

Highly correlated features: ['Car_Age', 'Insurance_Valid Until [date]', 'Registration Certificate_Not Available']


In [55]:
# reove the features from both training and testing
X_train_reduced = X_train_scaled.drop(columns=high_corr)
X_test_reduced = X_test_scaled.drop(columns=high_corr)

In [56]:
# check dimensions
print("Before correlation removal:", X_train_scaled.shape)
print("After correlation removal:", X_train_reduced.shape)

Before correlation removal: (2000, 912)
After correlation removal: (2000, 909)


### 5.2. Principal Component Analysis (PCA)

In [57]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.95) # means PCA keeps enough components to explain 95% of the variance

X_train_pca = pca.fit_transform(X_train_reduced)
X_test_pca = pca.transform(X_test_reduced)

In [58]:
print("Original features:", X_train_reduced.shape[1])
print("PCA components:", X_train_pca.shape[1])
print("Variance retained:", pca.explained_variance_ratio_.sum())

Original features: 909
PCA components: 660
Variance retained: 0.9508353119034939


- NB: PCA should be fitted only on X_train, which is why we use pca.fit_transform(X_train_reduced) and only pca.transform(X_test_reduced).

# **Modelling and Evaluation**

In [59]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

models = {
    "Linear Regression": LinearRegression(),
    "Ridge": Ridge(),
    "Lasso": Lasso(),
    "Random Forest": RandomForestRegressor(
        n_estimators=100, random_state=42
    ),
    "Gradient Boosting": GradientBoostingRegressor(
        random_state=42
    )
}

In [60]:
# train and compare them
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

for name, model in models.items():
    model.fit(X_train_pca, y_train)
    
    predictions = model.predict(X_test_pca)
    
    mae = mean_absolute_error(y_test, predictions)
    rmse = np.sqrt(mean_squared_error(y_test, predictions))
    r2 = r2_score(y_test, predictions)
    
    print(name)
    print("MAE:", mae)
    print("RMSE:", rmse)
    print("R²:", r2)
    print()

Linear Regression
MAE: 196488.55365548152
RMSE: 226900.00585431332
R²: 0.07098413261576342

Ridge
MAE: 196491.14249685415
RMSE: 226901.91712400562
R²: 0.07096848160383318

Lasso
MAE: 196488.60336408092
RMSE: 226899.90794282535
R²: 0.07098493439014808

Random Forest
MAE: 163813.05389999997
RMSE: 199881.19957172696
R²: 0.27906184320409344

Gradient Boosting
MAE: 167671.12216055577
RMSE: 202428.0353557879
R²: 0.2605727733785542



In [61]:
results = []

for name, model in models.items():
    model.fit(X_train_pca, y_train)
    predictions = model.predict(X_test_pca)

    results.append({
        'Model': name,
        'MAE': mean_absolute_error(y_test, predictions),
        'RMSE': np.sqrt(mean_squared_error(y_test, predictions)),
        'R²': r2_score(y_test, predictions)
    })

results_df = pd.DataFrame(results)

# Sort by MAE (lowest = best)
results_df = results_df.sort_values('MAE')

print(results_df)

               Model            MAE           RMSE        R²
3      Random Forest  163813.053900  199881.199572  0.279062
4  Gradient Boosting  167671.122161  202428.035356  0.260573
0  Linear Regression  196488.553655  226900.005854  0.070984
2              Lasso  196488.603364  226899.907943  0.070985
1              Ridge  196491.142497  226901.917124  0.070968


### Insights
- used MAE as primary metric since it tells the average amount by which the predicted car price differs from the actual price, in the same currency units as the target.
- Use RMSE as a secondary metric because it penalizes large prediction errors more heavily, and R² to show how much variation in car prices the model explains.
- Random Forest performed the best since it had the lowest MAE.

# **Conclusion**

## 7.1 Summary

- This project developed a machine learning model for predicting used-car prices. Feature engineering improved the dataset by creating useful features such as Car Age and price per mile and transforming the Accessories feature into individual binary indicators. 
- Missing categorical values were handled using "Unknown", while categorical variables were encoded using one-hot encoding and numerical features were standardized using StandardScaler.

- Dimensionality reduction was performed through correlation analysis and Principal Component Analysis (PCA). Highly correlated features were identified and removed, while PCA was then used to reduce the feature space while retaining 95% of the variance.

- Several regression models were evaluated using MAE, RMSE and R². MAE was used as the primary metric because it provides an easily interpretable measure of the average prediction error in the same units as car price. Random Forest performed best, achieving an MAE of approximately 163,813, an RMSE of approximately 199,881 and an R² of approximately 0.279. This indicates that Random Forest captured more of the relationships between the vehicle characteristics and its price than the other tested models.

- Although Random Forest achieved the best performance among the models tested, its R² value of 0.279 indicates that the model explains only approximately 27.9% of the variation in used-car prices. This suggests that important factors influencing car prices may not be captured in the available dataset. Additional features such as vehicle condition, engine capacity, location, service history, and market demand could potentially improve the model's predictive performance. Further hyperparameter tuning and testing of advanced boosting algorithms could also be explored.


## 7.2 Future Work
- Future improvements could include creating additional features such as vehicle mileage per year, vehicle age groups and interactions between vehicle characteristics. Additional data such as vehicle condition, location, engine specifications, and market demand could also improve prediction accuracy.

- Further work could explore dimensionality-reduction techniques such as feature selection and alternative PCA configurations. More advanced models, including XGBoost, LightGBM or tuned Random Forest models, could also be tested to potentially improve predictive performance.